File: learning-platform/src/services/progressService.js
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8080/api';

class ProgressService {
  // Get user's progress for a specific course
  async getUserCourseProgress(userId, courseId) {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/progress/user/${userId}/course/${courseId}`
      );
      return response.data;
    } catch (error) {
      console.error('Error fetching progress:', error);
      throw error;
    }
  }

  // Get all progress for a user
  async getAllUserProgress(userId) {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/progress/user/${userId}`
      );
      return response.data;
    } catch (error) {
      console.error('Error fetching all progress:', error);
      return [];
    }
  }

  // Mark lesson as complete
  async markLessonComplete(userId, courseId, lessonId) {
    try {
      const response = await axios.post(
        `${API_BASE_URL}/progress/update`,
        {
          userId,
          courseId,
          lessonId
        }
      );
      return response.data;
    } catch (error) {
      console.error('Error marking lesson complete:', error);
      throw error;
    }
  }

  // Mark lesson as incomplete (undo)
  async markLessonIncomplete(userId, courseId, lessonId) {
    try {
      const response = await axios.post(
        `${API_BASE_URL}/progress/undo`,
        {
          userId,
          courseId,
          lessonId
        }
      );
      return response.data;
    } catch (error) {
      console.error('Error marking lesson incomplete:', error);
      throw error;
    }
  }

  // Get progress summary (dashboard stats)
  async getProgressSummary(userId) {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/progress/user/${userId}/summary`
      );
      return response.data;
    } catch (error) {
      console.error('Error fetching progress summary:', error);
      return {
        totalCourses: 0,
        completedCourses: 0,
        totalLessons: 0,
        completedLessons: 0,
        overallProgress: 0
      };
    }
  }

  // Calculate progress percentage for a course
  calculateProgressPercentage(completedLessons, totalLessons) {
    if (totalLessons === 0) return 0;
    return Math.round((completedLessons / totalLessons) * 100);
  }
}

export default new ProgressService();

File: Forge/Forge/src/main/java/com/learnforge/service/ProgressService.java
package com.learnforge.service;

import com.learnforge.entity.Course;
import com.learnforge.entity.Lesson;
import com.learnforge.entity.Progress;
import com.learnforge.entity.User;
import com.learnforge.repository.*;
import jakarta.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.*;

@Service
public class ProgressService {

    @Autowired
    private ProgressRepository progressRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private CourseRepository courseRepository;

    @Autowired
    private LessonRepository lessonRepository;

    @Autowired
    private UserStreakService streakService;  // ✅ Fixed name

    // ✅ Get user progress for a course
    public Progress getUserCourseProgress(String userId, Long courseId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        Course course = courseRepository.findById(courseId)
                .orElseThrow(() -> new RuntimeException("Course not found"));

        return progressRepository.findByUserAndCourse(user, course)
                .orElseGet(() -> createNewProgress(user, course));
    }

    private Progress createNewProgress(User user, Course course) {
        Progress progress = new Progress();
        progress.setUser(user);
        progress.setCourse(course);
        progress.setCompletedLessons(new HashSet<>());
        progress.setCreatedAt(LocalDateTime.now());
        progress.setUpdatedAt(LocalDateTime.now());
        return progressRepository.save(progress);
    }

    public List<Progress> getAllUserProgress(String userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        return progressRepository.findByUser(user);
    }

    // ✅ Main method to mark lesson complete
    @Transactional
    public Progress markLessonComplete(String userId, Long courseId, Integer lessonId) {
        System.out.println("✅ Marking lesson " + lessonId + " complete for user " + userId);

        Progress progress = getUserCourseProgress(userId, courseId);

        if (progress.getCompletedLessons() == null) {
            progress.setCompletedLessons(new HashSet<>());
        }

        progress.getCompletedLessons().add(lessonId);
        progress.setUpdatedAt(LocalDateTime.now());

        // ✅ Update streak when lesson is completed
        streakService.updateStreakOnLessonComplete(userId);

        return progressRepository.save(progress);  // ✅ Fixed: lowercase 'p'
    }

    @Transactional
    public Progress markLessonIncomplete(String userId, Long courseId, Integer lessonId) {
        System.out.println("❌ Marking lesson " + lessonId + " incomplete for user " + userId);

        Progress progress = getUserCourseProgress(userId, courseId);

        if (progress.getCompletedLessons() != null) {
            progress.getCompletedLessons().remove(lessonId);
            progress.setUpdatedAt(LocalDateTime.now());
        }

        return progressRepository.save(progress);  // ✅ Fixed: lowercase 'p'
    }

    public Map<String, Object> getProgressSummary(String userId) {
        List<Progress> progressList = getAllUserProgress(userId);

        int totalCourses = progressList.size();
        int completedCourses = 0;
        int totalLessons = 0;
        int completedLessons = 0;

        for (Progress progress : progressList) {
            Course course = progress.getCourse();
            List<Lesson> lessons = lessonRepository.findByCourse(course);

            int courseLessonCount = lessons.size();
            int courseCompletedCount = progress.getCompletedLessons() != null
                    ? progress.getCompletedLessons().size() : 0;

            totalLessons += courseLessonCount;
            completedLessons += courseCompletedCount;

            if (courseLessonCount > 0 && courseCompletedCount == courseLessonCount) {
                completedCourses++;
            }
        }

        Map<String, Object> summary = new HashMap<>();
        summary.put("totalCourses", totalCourses);
        summary.put("completedCourses", completedCourses);
        summary.put("totalLessons", totalLessons);
        summary.put("completedLessons", completedLessons);
        summary.put("overallProgress", totalLessons > 0 ? (completedLessons * 100.0 / totalLessons) : 0);

        return summary;
    }
}
